### PR TITLE
Nicos/audionodegraph

### PR DESCRIPTION
**Pull request by nicolasgustafsson**
Implement node graphs w/ editor
Implement audio mixing and sound effect node graphs

### CODE REVIEW

File: WonderMake/Audio/AudioManager.cpp [hunk 1]
```diff
@@ -28,25 +31,48 @@ AudioManager::~AudioManager()
...
 }
 
 void AudioManager::PlayAudio(ResourceProxy<AudioFile> aAudioFileToPlay)
 {
-	auto handle = myBus.play(aAudioFileToPlay->GetSource());
+	auto handle = myBusses["Gameplay"].play(aAudioFileToPlay->GetSource());
 	myCurrentlyPlayingAudioFiles.insert({ handle, aAudioFileToPlay });
 }
 
+SoLoud::Bus& AudioManager::GetBus(std::string aBusName)
```
Comment (Devilhar) [inline WonderMake/Audio/AudioManager.cpp:66]: const ref

File: WonderMake/Audio/AudioMixingNodeGraph.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#include "pch.h"
+#include "AudioMixingNodeGraph.h"
+#include "Audio/AudioNodeTypes.h"
+
+AudioMixingNodeGraph::AudioMixingNodeGraph(std::filesystem::path aPath)
+	:  NodeGraph(aPath)
```
Comment (Devilhar) [inline WonderMake/Audio/AudioMixingNodeGraph.cpp:6]: NodeGraph(std::move(aPath))

File: WonderMake/Audio/AudioNodeTypes.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		busPointer->setFilter(3, nullptr);
+
+		aNode.SetOutput<SoLoud::Bus*>(0, busPointer);
+	}
+
+	void SAudioMixNode::ExecuteNodeLeftToRight(SNode& aNode)
+	{
+		std::any& audioMix = aNode.NodeData["AudioMix"];
+		SoLoud::Bus* busPointer = nullptr;
+		busPointer = &std::any_cast<SoLoud::Bus&>(audioMix);
```
Comment (Devilhar) [inline WonderMake/Audio/AudioNodeTypes.cpp:47]: SoLoud::Bus& busPointer = std::any_cast<SoLoud::Bus&>(audioMix);

File: WonderMake/NodeGraph/SlotCustomization/SlotSerialization.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+		TSlotType DeserializeInput(const json& aJson)
+		{
+			return aJson["Value"].get<TSlotType>();
+		}
+
+		template<>
+		inline std::filesystem::path DeserializeInput<std::filesystem::path>(const json& aJson)
+			{
+				return std::filesystem::path(aJson["Value"].get<std::string>());
+			}
```
Comment (Devilhar) [inline WonderMake/NodeGraph/SlotCustomization/SlotSerialization.h:47]: Fix indentation.

File: WonderMakeTests/WonderMakeTests.vcxproj [hunk 5]
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+?�<?xml version="1.0" encoding="utf-8"?>
```
Comment (Devilhar) [inline WonderMakeTests/WonderMakeTests.vcxproj:1]: Remove added characters.

File: WonderMake/NodeGraph/NodeGraph.cpp [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+	}
+}
+
+void NodeGraph::KillNodesThatWantToDie()
+{
+	plf::colony<SNode>::colony_iterator<false> it = myNodes.begin();
+
+	while (it != myNodes.end())
+	{
+		//[Nicos]: Root nodes can't die even if they want to
```
[the remark this answers is not in the document]
Comment (nicolasgustafsson) [reply WonderMake/NodeGraph/NodeGraph.cpp:114]: How do you know this stuff

File: WonderMake/Audio/AudioManager.cpp (hunk 1 above)
Comment (nicolasgustafsson) [reply WonderMake/Audio/AudioManager.cpp]: Fixed

File: WonderMake/Audio/AudioMixingNodeGraph.cpp (hunk 2 above)
Comment (nicolasgustafsson) [reply WonderMake/Audio/AudioMixingNodeGraph.cpp]: Fixed

File: WonderMake/Audio/AudioNodeTypes.cpp (hunk 3 above)
Comment (nicolasgustafsson) [reply WonderMake/Audio/AudioNodeTypes.cpp]: Fixed

File: WonderMake/Audio/AudioManager.cpp [hunk 7]
```diff
@@ -27,25 +30,41 @@ AudioManager::~AudioManager()
...
+	auto it = std::find_if(mySoundEffects.begin(), mySoundEffects.end(), [aAudioPath](const SoundEffectNodeGraph& aSoundEffect) { return aSoundEffect.GetName() == aAudioPath.string(); });
+
+	if (it == mySoundEffects.end())
+		it = mySoundEffects.emplace(aAudioPath);
 
-	if (audioFile.IsValid())
-		PlayAudio(audioFile);
-	else
-		myQueuedAudioFiles.insert({ audioFile });
+	(*it).Execute();
```
[the remark this answers is not in the document]
Comment (nicolasgustafsson) [reply WonderMake/Audio/AudioManager.cpp]: Fixed

File: WonderMake/Audio/AudioNodeTypes.cpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+			SoLoud::EchoFilter* filter = nullptr;
+
+			if (anyFilter.has_value())
+			{
+				filter = &std::any_cast<SoLoud::EchoFilter&>(anyFilter);
+			}
+			else
+			{
+				filter = &anyFilter.emplace<SoLoud::EchoFilter>();
+			}
```
[the remark this answers is not in the document]
Comment (nicolasgustafsson) [reply WonderMake/Audio/AudioNodeTypes.cpp]: Fixed

File: WonderMake/Imgui/Canvas.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		{
+			if (io.MouseWheel != 0)
+			{
+				ImVec2 mouseRel = ImVec2{ ImGui::GetMousePos().x - ImGui::GetWindowPos().x, ImGui::GetMousePos().y - ImGui::GetWindowPos().y };
+				float prevZoom = aCanvasState->ZoomLevel;
+				aCanvasState->ZoomLevel = std::clamp(aCanvasState->ZoomLevel + io.MouseWheel * aCanvasState->ZoomLevel / 16.f, 0.3f, 3.f);
+				float zoomFactor = (prevZoom - aCanvasState->ZoomLevel) / prevZoom;
+
+				aCanvasState->Offset.x += ImVec2((mouseRel.x - aCanvasState->Offset.x) * zoomFactor, (mouseRel.y - aCanvasState->Offset.y) * zoomFactor).x;
+				aCanvasState->Offset.y += ImVec2((mouseRel.x - aCanvasState->Offset.x) * zoomFactor, (mouseRel.y - aCanvasState->Offset.y) * zoomFactor).y;
```
[the remark this answers is not in the document]
Comment (nicolasgustafsson) [reply WonderMake/Imgui/Canvas.cpp:46]: Yes we should!

File: WonderMake/NodeGraph/SlotCustomization/SlotSerialization.h (hunk 4 above)
Comment (nicolasgustafsson) [reply WonderMake/NodeGraph/SlotCustomization/SlotSerialization.h]: Fixed

File: WonderMakeTests/WonderMakeTests.vcxproj (hunk 5 above)
Comment (nicolasgustafsson) [reply WonderMakeTests/WonderMakeTests.vcxproj]: Fixed

File: WonderMake/Audio/AudioMixingNodeGraph.cpp (hunk 2 above)
Comment (Devilhar) [reply WonderMake/Audio/AudioMixingNodeGraph.cpp]: Nope

Comment (nicolasgustafsson) [reply WonderMake/Audio/AudioMixingNodeGraph.cpp]: Yep

**Pull request merged**